### ISSUE TITLE
关于Encodec压缩量化的label 使用dictionary.encode_line编码的问题

### ISSUE THREAD

**Issue by xujinchang** [yizhilll/MERT#8]
你好，我使用了prepare_codecs_from_manifest.py这个脚本处理music4all中的一些音乐文件生成了8个txt, 这些值在mert_model里面直接load成dictionary，我打印断点看dictionary这个类下面的方法，line, field = line.rstrip().rsplit(" ", 1)， count = int(field), word = line 是下图的值，这个dictionary在训练中又会对每一行的值重新编码，我得到的label是全3的值，想知道如果都是一样的label，怎么算ce loss呢
![image](https://github.com/yizhilll/MERT/assets/23400146/bc18c1c0-3870-4e61-93ea-8381ec25b7fa)

**Comment by yizhilll** [comment]
I translated the inquiry to maintain the consistency language use of the repo:

> Hello, I used the script prepare_codecs_from_manifest.py to process some music files from "music4all" and generated 8 txt files. These values are directly loaded as a dictionary in the mert_model. I printed a breakpoint to inspect the methods under the dictionary class, where line, field = line.rstrip().rsplit(" ", 1), count = int(field), and word = line represent the values in the image below. This dictionary will re-encode each line's value during training. The label I obtained is consistently the value 3. I'm curious about how to calculate the CE (cross-entropy) loss when all labels are identical?

Regarding to your question, although the extracted codes are the same, their corresponding audio segements are different (which can be calculated according to the frequency 75Hz of the EnCodec model). For example, the losses corresponding to audio segments at [..., t_i, t_i+1, t_i+2, t_i+3, ...] are calculated seperately. 

Specifically, in the practice of our codes, there are randomly initialised embeddings corresponding to all the discrete labels, i.e. the 8*1024 codecs, K_MFCC in the K-means_MFCC, or K_logmel in the K-means_logmel. If you look back to the [code base](https://github.com/yizhilll/MERT/blob/b2fb3cca5f3e2054617b879f55292fa21517e1c2/mert_fairseq/models/mert/mert_model.py#L953), you’ll find that the NCE loss is calculated between such label-embeddings [..., y_i, y_i+1, y_i+2, y_i+3, ...] and the output of the model [..., o_i, o_i+1, o_i+2, o_i+3, ...] for diferent codebooks individually and then aggregated. 
In the end, all the logics are summed up and returned.

**Comment by xujinchang** [comment]
Thanks for your quick reply. 

I have run the scripts provided by hubert using k-means labels and check the training data. For k-means,  there are two files, one is "train.km" and the other is "dict.km.txt".

For codecs, I think that prepare_codecs_from_manifest.py just provides the files like "train.km". I should count the numbers in the "train.km" and produce the "dict.km.txt" by myself.  

In my case, the values of labels are all 3 because the encode_line returns the unk_index which is 3. I should using the "dict.km.txt" to inialize the Dictionary class instead of  "train.km".

That‘s right?

**Comment by yizhilll** [comment]
> Thanks for your quick reply.
> 
> I have run the scripts provided by hubert using k-means labels and check the training data. For k-means, there are two files, one is "train.km" and the other is "dict.km.txt".
> 
> For codecs, I think that prepare_codecs_from_manifest.py just provides the files like "train.km". I should count the numbers in the "train.km" and produce the "dict.km.txt" by myself.
> 
> In my case, the values of labels are all 3 because the encode_line returns the unk_index which is 3. I should using the "dict.km.txt" to inialize the Dictionary class instead of "train.km".
> 
> That‘s right?

Yes! That is correct.

**Comment by xujinchang** [comment]
I have trained MERT-95M using music4all data and the loss descend slowly, Is it normal?
![image](https://github.com/yizhilll/MERT/assets/23400146/3cb178ae-2bfb-4d63-b7b5-f9f4d60dc48b)
![image](https://github.com/yizhilll/MERT/assets/23400146/0d7b1b42-8a09-431e-9d83-9cb11d431d75)

By the way, I make "dict.encodec_0.txt" ... "dict.encodec_7.txt"  are all the same,  
`codebook_size=1024

for x in $(seq 0 $((codebook_size - 1))); do
  echo "$x 1"
done > dict.km.txt

for i in {0..7}
do
  cp dict.km.txt dict.encodec_${i}.txt
done
~                       `

self.num_classes = [len(d) for d in dictionaries] The num_classes is (1024+4)*8. Are these parameters correct?

**Comment by xujinchang** [comment]
I have found the targets in the cross_entropy function are all zero values, is it a bug or feature? Can u show the examples of the dict_encodec_{i}.txt？ I think the 2nd column is the count of each symbol but its not actually used and I can just put any number.
![image](https://github.com/yizhilll/MERT/assets/23400146/97c28116-0e2b-47a1-ad18-469ba87d543f)
![image](https://github.com/yizhilll/MERT/assets/23400146/f40298fa-ec70-4790-b9fc-f69d2e23f44e)

**Comment by yizhilll** [comment]
> I have found the targets in the cross_entropy function are all zero values, is it a bug or feature? Can u show the examples of the dict_encodec_{i}.txt？ I think the 2nd column is the count of each symbol but its not actually used and I can just put any number. ![image](https://user-images.githubusercontent.com/23400146/276806719-97c28116-0e2b-47a1-ad18-469ba87d543f.png) ![image](https://user-images.githubusercontent.com/23400146/276806796-f40298fa-ec70-4790-b9fc-f69d2e23f44e.png)

The dummy dictionary construction codes are from [HuBERT](https://github.com/facebookresearch/fairseq/tree/main/examples/hubert/simple_kmeans#create-a-dummy-dict). Could you confirm your dictionary are the same?
